File: Leet 2013/solution.cs
public class DetectSquares
{
    struct Point
    {
        public int X { get; set; }
        public int Y { get; set; }
    }

    private readonly Dictionary<Point, int> _points = [];

    public DetectSquares()
    {

    }

    public void Add(int[] point)
    {
        var pt = new Point
        {
            X = point[0],
            Y = point[1]
        };

        _points[pt] = GetValueFromPoints(pt) + 1;
    }

    public int Count(int[] point)
    {
        var pt = new Point
        {
            X = point[0],
            Y = point[1]
        };

        int result = 0;
        foreach (var (key, _) in _points)
        {
            // Skip points that are on the same row or column as the given point
            if (pt.X == key.X || pt.Y == key.Y)
            {
                continue;
            }

            // Check if the distance between the given point and the current point is equal in both x and y directions
            int dx = Math.Abs(key.X - pt.X);
            int dy = Math.Abs(key.Y - pt.Y);
            if (dx != dy)
            {
                continue;
            }

            // Calculate the number of squares that can be formed with the given point and the current point
            var a = GetValueFromPoints(new Point { X = key.X, Y = pt.Y });
            var b = GetValueFromPoints(new Point { X = pt.X, Y = key.Y });
            var c = GetValueFromPoints(key);
            result += a * b * c;
        }

        return result;
    }

    // Helper method to get the value from the dictionary for a given point or return 0 if the point does not exist in the dictionary
    private int GetValueFromPoints(Point key)
    {
        if (_points.TryGetValue(key, out int value))
        {
            return value;
        }

        return 0;
    }
}

class Program
{
    private static void Main()
    {
        Console.WriteLine("2013. Detect Squares");

        DetectSquares detectSquares = new();

        detectSquares.Add([3, 10]);
        detectSquares.Add([11, 2]);
        detectSquares.Add([3, 2]);
        Console.WriteLine($"Count squares with [11, 10]: {detectSquares.Count([11, 10])}");
        Console.WriteLine($"Count squares with [14, 8]: {detectSquares.Count([14, 8])}");
        detectSquares.Add([11, 2]);
        Console.WriteLine($"Count squares with [11, 10]: {detectSquares.Count([11, 10])}");
    }
}

File: Leet 2013/solution.go
package main

import (
	"fmt"
	"math"
)

type Point struct {
	X int
	Y int
}

type DetectSquares struct {
	points map[Point]int
}

func Constructor() DetectSquares {
	return DetectSquares{
		points: make(map[Point]int),
	}
}

func (this *DetectSquares) Add(point []int) {
	this.points[Point{X: point[0], Y: point[1]}]++
}

func (this *DetectSquares) Count(point []int) int {
	pt := Point{X: point[0], Y: point[1]}
	var result int

	for key := range this.points {
		if pt.X == key.X || pt.Y == key.Y {
			continue
		}

		dx := math.Abs(float64(key.X - pt.X))
		dy := math.Abs(float64(key.Y - pt.Y))
		if dx != dy {
			continue
		}

		a := this.points[Point{X: key.X, Y: pt.Y}]
		b := this.points[Point{X: pt.X, Y: key.Y}]
		c := this.points[key]
		result += a * b * c
	}

	return result
}

func main() {
	fmt.Println("2013. Detect Squares")

	detectSquares := Constructor()

	detectSquares.Add([]int{3, 10})
	detectSquares.Add([]int{11, 2})
	detectSquares.Add([]int{3, 2})

	fmt.Printf("Count([11,10]) = %d\n", detectSquares.Count([]int{11, 10}))
	fmt.Printf("Count([14,8]) = %d\n", detectSquares.Count([]int{14, 8}))

	detectSquares.Add([]int{11, 2})

	fmt.Printf("Count([11,10]) = %d\n", detectSquares.Count([]int{11, 10}))

}
